### ISSUE TITLE
Operations on Columns of Differing Types

### ISSUE THREAD

**Issue by amin-nejad** [databricks/koalas#1819]
I just noticed this difference between `koalas` and `pandas` when it comes to performing operations on columns of different types e.g. `int` and `str`. Pandas will throw a `TypeError` whereas `koalas` won't and will actually return something. I'm sure this isn't a bug but I wondered if there's anywhere where this behaviour is documented that I can read about? The return types seem to be a little unpredictable. I have only looked at `str` and `int` so far but it would be great if there was more information on all the different types and why they return what they return. Also, is there a reason, we don't throw a `TypeError`? This would be a very quick check just based on the `dtype` of the column

### Koalas
```python
df = ks.DataFrame({'a':[1, 2, 3], 'b':['a', '2', 'c'], 'c':['d', 'e', 'f']})
df['add_ab'] = df['a'] + df['b']
df['sub_ab'] = df['a'] - df['b']
df['div_ab'] = df['a'] / df['b']
df['mul_ab'] = df['a'] * df['b']
df['add_bc'] = df['b'] + df['c']
df['sub_bc'] = df['b'] - df['c']
df['div_bc'] = df['b'] / df['c']
df['mul_bc'] = df['b'] * df['c']
print(df)
#    a  b  c  add_ab  sub_ab  div_ab  mul_ab add_bc  sub_bc    div_bc  mul_bc
# 0  1  a  d     NaN     NaN     inf     NaN     ad     NaN      None     NaN
# 1  2  2  e     4.0     0.0     1.0     4.0     2e     NaN  Infinity     NaN
# 2  3  c  f     NaN     NaN     inf     NaN     cf     NaN      None     NaN
```

- Interestingly, '2' as a string in column 'b' behaves as a number.
- div_ab results in `inf` whereas the other ab operations result in `NaN`
- div_bc this time results in `None` where the other operations (ignoring add) result in `NaN` again
    - except for row 2 which this time results in `Infinity`

Essentially, I'm not sure why the string version of 2 gets typecast to an integer. And why division results in behaviour that is different to the other operations. And why this is sometimes `None` and sometimes `inf`. And what the difference is between `inf` and `Infinity`.

### Pandas
```python
df = pd.DataFrame({'a':[1, 2, 3], 'b':['a', '2', 'c'], 'c':['d', 'e', 'f']})
df['mul_ab'] = df['a'] * df['b']
df['add_bc'] = df['b'] + df['c']
print(df)
#    a  b  c mul_ab add_bc
# 0  1  a  d      a     ad
# 1  2  2  e     22     2e
# 2  3  c  f    ccc     cf
```

Comparing to the `pandas` version. The two operations included above are the only two that do not result in a `TypeError`. Here the 2 remains as a string. And the allowed operations are as expected.

## Sidenote
```python
df = ks.DataFrame({'a':[1, 2, 3], 'b':['a', 2, 'c'], 'c':['d','e', 'f']})
```
throws an `ArrowTypeError` (the 2 in col 'b' is now an int rather than a string). Pandas however allows this Dataframe to be created. Furthermore, I'm not sure how to make sense of this given that the 2 gets effectively turned into an `int` anyway if given as a `str`.

## Summary

To summarise, I'm not sure what of the above are bugs (if any). But a detailed breakdown of how functions on columns of different types behave compared to their pandas counterparts would be much appreciated. 

koalas==1.2.0
pandas==1.0.5

**Comment by ueshin** [comment]
Some of the string operations seem bugs. Let us investigate it and submit a PR.

Regarding the sidenote, PySpark (and PyArrow) doesn't allow to have different data type values in a single column, so neither does Koalas. That's one of strict restrictions of Koalas.
